src/Parser/Core/Modules/Helpers/ApplyBuffFixer.js

```diff
@@ -0,0 +1,130 @@
+      this._combatantInfoEvents.push(event);
+    }
+  }
+
+  _buffsAppliedByPlayerId = {};
+
+  /**
+   * Some buffs like Bloodlust and Holy Avenger when they are applied pre-combat it doesn't show up as an `applybuff` event nor in the `combatantinfo` buffs array. It can still be detected by looking for a `removebuff` event, this uses that to detect it and then fabricates an `applybuff` event at the start of the log.
+   * @param {Array} events
+   * @returns {Array}
+   */
+  reorderEvents(events) {
+    const firstEventIndex = this._getFirstEventIndex(events);
+    const playersById = this.owner.playersById;
+
+    // region Buff event based detection
+    // This catches most relevant buffs and is most accurate. If a player is good at juggling certain buffs they can achieve 100% uptime, if that happens `removebuff` is never called, so we also check for other indicators that are just as reliable, such as `applybuffstack`, `removebuffstack` and `refreshbuff`.
+    for (let i = 0; i < events.length; i += 1) {
+      const event = events[i];
+      const targetId = event.targetID;
+
+      this._buffsAppliedByPlayerId[targetId] = this._buffsAppliedByPlayerId[targetId] || [];
+
+      if (event.type === 'applybuff') {
+        const spellId = event.ability.guid;
+        this._buffsAppliedByPlayerId[targetId].push(spellId);
+      }
+      if (['removebuff', 'applybuffstack', 'removebuffstack', 'refreshbuff'].includes(event.type)) {
+        const spellId = event.ability.guid;
+        if (this._buffsAppliedByPlayerId[targetId].includes(spellId)) {
+          // This buff has an `applybuff` event and so isn't broken :D
+          continue;
+        }
+
+        const fightDuration = formatDuration((event.timestamp - this.owner.fight.start_time) / 1000);
+        debug && console.warn(fightDuration, 'Found a buff on', ((playersById[targetId] && playersById[targetId].name) || '???'), 'that was applied before the pull:', event.ability.name, spellId, '! Fabricating an `applybuff` event so you don\'t have to do anything special to take this into account.');
+        const applybuff = {
+          // These are all the properties a normal `applybuff` event would have.
+          timestamp: events[firstEventIndex].timestamp,
+          type: 'applybuff',
+          ability: event.ability,
+          sourceID: event.sourceID,
+          sourceIsFriendly: event.sourceIsFriendly,
+          targetID: targetId,
+          targetIsFriendly: event.targetIsFriendly,
+          // Custom properties:
+          prepull: true,
+          __fabricated: true,
+        };
+        events.splice(firstEventIndex, 0, applybuff);
+        // It shouldn't happen twice, but better be safe than sorry.
+        this._buffsAppliedByPlayerId[targetId].push(spellId);
+      }
+    }
+    // endregion
+
+    // region `combatantinfo` based detection
+    // This catches buffs that never drop, such as Flasks and more importantly.
```

Comment:
and more importantly WHAT? I'm dying to know!

Reply:
I couldn't think of anything. The check is kinda useless.

Reply:
Actually I was thinking Atonements on other players. And beacons.

Reply:
yeah I figured you had something in mind, was mostly noting that that comment seemed to trail off mid sentence :)